EasyFullControl fleet_adapter_mir and cart delivery PerformAction plugin

xiyuoh opened the pull request:
This PR:
- Added note that marks MiRFleet adapter as inactive
- Switches MiR fleet adapter to use the EasyFullControl API
- Creates standard RMF missions from the fleet adapter via MiR's REST API calls so that users don't have to manually add them on the hardware. This also allows us to automatically create more missions on the MiR required by RMF in the future. Mission definitions can be found in `rmf_missions.json`.
- Various improvements on the fleet adapter, such as using mission queue IDs to monitor mission/task completion, better localizing process when switching maps to a new level, and an expanded list of RMF missions that help robots get to a MiR position more accurately instead of only relying on coordinates
- Adds a plugin system for MiR custom actions. Plugins can be defined and configured inside `mir_config.yaml`
- Adds a CartDelivery perform action plugin that helps users to integrate delivery tasks with RMF. Users will need to create their own pickup and dropoff missions and specify the mission names inside `mir_config.yaml`.  This plugin includes a task script for users to submit `dispatch_pickup` tasks. Remaining cart delivery related missions required by RMF can be automatically created on hardware during runtime by parsing `rmf_cart_missions.json`.

One major flaw in this fleet adapter is the `is_charging` function inside `robot_adapter_mir.py`. As far as I know and after checking with vendors, there doesn't seem to be a MiR API that we can use to check whether the robot has begun charging. The MiR status provides a `mode_id` and `state_id`, currently there's no known state/mode for charging. Will experiment with hardware to verify this.

Tested on hardware with MiR200 and MiR250:
- [x]  Standard RMF  mission creation (rmf_missions.json)
- [x] Fleet adapter able to launch and run
- [x] MiR able to perform patrol tasks and map switch
- [x] RMF cart delivery mission creation (rmf_cart_missions.json)
- [x] MiR able to perform delivery tasks

xiyuoh commented:
Update from MiR vendor: no available indication of when the robot has started charging. There is no existing `state_id` or `mode_id` meant for that data. Using mission status text is currently the only way to verify whether charging has started.

mxgrey commented on fleet_adapter_mir_tasks/fleet_adapter_mir_tasks/dispatch_pickup.py:
```diff
@@ -0,0 +1,208 @@
+#!/usr/bin/env python3
```
I'm worried that calling this file `dispatch_pickup.py` might lead users to think it only does a pickup and not a drop-off, when in reality it does an entire delivery. I assume you were trying to avoid using the same name as the `dispatch_delivery.py` file in `rmf_demos_tasks`, but we shouldn't use a different name if it comes at the cost of clarity.

This file will be namespaced inside of `fleet_adapter_mir_tasks` so there won't be a conflict if it has the same name as `dispatch_delivery` from `rmf_demos_tasks`. Users will run it like this:
```
$ ros2 run fleet_adapter_mir_tasks dispatch_delivery ...
```

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/fleet_adapter_mir.py:
```diff
@@ -286,18 +173,20 @@ def main(argv=sys.argv):
                         help="Input config.yaml file to process")
     parser.add_argument("-n", "--nav_graph", type=str, required=True,
                     help="Path to the nav_graph for this fleet adapter")
+    parser.add_argument("-a", "--actions", type=str, required=False, default='',
```
As we move towards a modular plugin design, I think it would make sense to let this be a list of action files instead of only one action file. Then users can add their own custom actions on top of the ones we provide by writing their own additional .json files.

xiyuoh replied:
This `actions` argument takes in the base rmf missions (path to `rmf_missions.json` [here](https://github.com/osrf/fleet_adapter_mir/tree/xiyu/efc_fleet_adapter_mir/missions)) required for default tasks such as patrol, localize and charging. Any perform action-related missions will be parsed in from the config as a [field under each plugin](https://github.com/osrf/fleet_adapter_mir/blob/xiyu/efc_fleet_adapter_mir/configs/mir_config.yaml#L111).

I realized naming it `actions` is extremely misleading. Actions have a different meaning associated with MiR missions (they are the building blocks of the missions). I've changed it to `rmf_missions`.

mxgrey commented on fleet_adapter_mir/README.md:
```diff
@@ -41,8 +41,72 @@ Alternatively, if you want to run everything with full capabilities, (though not
+- `rmf_dock_to_cart`: Docks robot under the cart
+- `rmf_exit_lot`: Calls the robot to exit from under the cart
+- `rmf_update_footprint`: Updates the robot footprint
+They are defined and stored in the `rmf_cart_missions.json` file and do not require any further configuration.
```
You need an extra newline before this line, otherwise the markdown puts this sentence into the preceding bullet point.

mxgrey commented on configs/mir_config.yaml:
```diff
@@ -85,17 +69,45 @@ rmf_fleet:
+
+plugins:
+  rmf_cart_delivery:
+    actions: ["delivery_pickup", "delivery_dropoff"]
```
To move towards a true plugin-based system, we can add a `module` key to this dictionary that lets users say what module to import the plugin from, and the class name to instantiate as an action, e.g.:
```
rmf_cart_delivery:
    module: fleet_adapter_mir.rmf_cart_delivery
    class: CartDelivery
```

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py:
```diff
@@ -0,0 +1,491 @@
+                name=name,
+                mir_api=self.api,
+                update_handle=self.update_handle,
+                actions=plugin_config['rmf_cart_delivery']['actions'],
```
Instead of passing this and the next two arguments in as all separate arguments, why not just pass in the `plugin_config` and parse these fields out later?

mxgrey reviewed (changes requested):
There are a lot of excellent designs in this PR. Using the json files to define the MiR missions, and then loading and configuring them at startup is fantastic.

This PR presents a huge improvement over the previous state of the fleet adapter. There's only one problem: The current approach to plugin support does not actually support plugins. Instead the only way a downstream user can integrate or customize their own actions is to fork the source code of this repository and edit it.

At first I considered accepting this as a temporary limitation, merging the PR as-is, and then improving the situation later. However, I'm concerned that early adopters of these changes will end up being punished with a diverging code base after forking as needed. That makes it difficult for them to take in improvements that we're going to make in the future and it also discourages downstream users from contributing anything upwards.

I also feel that the gap to support plugins from where this PR is right now is actually not very large. The key thing will be understanding how to use [`importlib`](https://stackoverflow.com/a/41678146) to load a module and then `getattr` to instantiate a class from that module. I left a few tips in my inline comments, but let's talk about this more in person when we have a chance.

xiyuoh commented on fleet_adapter_mir/README.md:
```diff
@@ -41,8 +41,72 @@ Alternatively, if you want to run everything with full capabilities, (though not
+
+**MiR positions**
+
+For more accurate robot maneuver, you may wish to send the MiR to a Robot Position instead of using coordinates. For such waypoints, you can provide a `dock_name` with the `rmf_move_to_position` mission specified:
```
You are right; removed and replaced with an explanation about identical position/waypoint names.

xiyuoh commented:
After our previous discussion, I've refactored the way we are creating `MirAction` objects in this new plugin system. They are now
- Only created when a task comes in with the performable action, and destroyed when the task is completed
- Created via an `ActionFactory`, and is tracked in the RobotAdapter under `self.current_action`
- More accurate plugin system with the use of `importlib` to import the relevant modules

xiyuoh commented:
This PR is ready for its next round of review.

The latest update refactors the way the `CartDelivery` action is designed. As discussed above and offline, we want a true plugin system where users do not have to modify and implement their own code in the library. However, for this perform action, we would need to know the user-specific API/logic to check whether the robot's latch is open, whether it is under a cart, etc.

With the current change, users will instead have to implement their own `CartDetection` module and provide it to the `CartDelivery` plugin by filling up the `cart_detection_module` field in the plugin config. An example of how a cart detection module can be implemented is demonstrated in the `rmf_cart_detection.py` script.

Other changes include:
- Adding a `disconnect_mutex` to the fleet adapter to ensure that new navigation/stop commands won't be called if a robot is disconnected (i.e. robot status cannot be retrieved), and triggers a replan when the robot reconnects.
- Allow robots to use MiR positions configured in the fleet config under `lift_positions` to perform localize missions. This means instead of localizing to a new map using RMF coordinates, the robot will localize to the provided MiR position on the new map, ensuring more accurate map switching.
  - If the position is not provided in the fleet config, the fleet adapter will create its own MiR position on the robot for this localize mission.
  - If the position is provided in the config but not found on the robot, the fleet adapter will create a MiR position on the robot with the name provided and use it for the localize mission.
- Remove `position_put` logic for `get_position_guid`. This is being used when localizing to a new map. This is to prevent the fleet adapter from overwriting any MiR position coordinates stored on the robot.
  - This was initially added to ensure that the MiR coordinates are updated according to the RMF waypoint. However, as localize missions are usually called when robots are inside the lift with little space for error, it is likely that the new MiR coordinates converted from RMF waypoint become "unreachable" for the robot, and the robot gets stuck being unable to reach its goal waypoint. Since we allow users to configure the MiR position used in lifts now, we should not update the provided position anymore and only use coordinates converted from RMF to create the MiR position if it doesn't exist on the robot.
- Added retry mechanism for `navigate` and `go_to_known_position` if the posted mission was not received by the robot
- Minor fixes, codestyle and cleanup

mxgrey commented on fleet_adapter_mir/README.md:
```diff
@@ -41,8 +41,70 @@ Alternatively, if you want to run everything with full capabilities, (though not
 
-An example configuration file, `mir_config.yaml` has been provided. It has been generously commented, and in the cases where it has not, the parameter names are self-explanatory enough.
\ No newline at end of file
+An example configuration file, `mir_config.yaml` has been provided. It has been generously commented, and in the cases where it has not, the parameter names are self-explanatory enough.
```
From reading the Python code implementation, it looks like the localization mission in particular has quite a few requirements around how it needs to be constructed which a user would only know from reading and understanding the Python code or from repeatedly trying to run the fleet adapter and looking at what assertions fail.

I think it would be best if we can document a rough sketch of what each basic mission type is supposed to look like. There should be enough information in the documentation for users to recreate the missions themselves in the MiR UI without looking at our Python code.

You could include that documentation here in this README, but it may be cleaner to create a new `docs` folder and put a file in there that the README can link to.

> in the cases where it has not, the parameter names are self-explanatory enough

The wording here may come off a bit aggressive. I recommend tweaking this to "in the cases where it has not, the parameter names are meant to be self-explanatory". Otherwise it may sound like we're blaming the reader if they don't understand the parameter names, but things that seem self-evident to us will not be self-evident to everyone.

mxgrey replied:
I'm noticing that some of the missions are created automatically, so we should also document what missions will be created automatically and what those will look like to the user so they don't get alarmed when they see new missions pop up in the MiR after using this fleet adapter.

xiyuoh replied:
I agree the mission organization is a little messy right now, I'll take the `docs` folder suggestion and update accordingly.

> The wording here may come off a bit aggressive. I recommend tweaking this to "in the cases where it has not, the parameter names are meant to be self-explanatory". Otherwise it may sound like we're blaming the reader if they don't understand the parameter names, but things that seem self-evident to us will not be self-evident to everyone.

Looks like this line is from the current package README, but I'll take note and modify the language. Thanks for highlighting!

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py:
```diff
@@ -0,0 +1,767 @@
+        retry_count = 10
+        while count < retry_count and not mission_queue_id:
+            if (navigation_mission_identifier !=
+                    mission_handle.execution.identifier):
```
You should use [`mission_handle.activity`](https://github.com/open-rmf/fleet_adapter_mir/blob/fc2e76993df4cffb2bc4b7e3555fc2aa2380750c/fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py#L62) here, just in case `mission_handle.execution` ever gets replaced with a `None` value.

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py:
```diff
@@ -0,0 +1,767 @@
+                    mission_handle.execution.identifier):
+                self.node.get_logger().info(
+                    f'[{self.name}] MissionHandle has changed to '
+                    f'{mission_handle.execution.identifier}, interrupting '
```
Same here, `mission_handle.activity`.

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py:
```diff
@@ -0,0 +1,767 @@
+
+        if (mission_queue_id is None and
+                navigation_mission_identifier !=
+                mission_handle.execution.identifier):
```
Same here, `mission_handle.activity`.

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py:
```diff
@@ -0,0 +1,767 @@
+                f'[{self.name}] Failed to request robot to move to known '
+                f'position. Navigation step {navigation_mission_identifier} '
+                f'interrupted by new mission '
+                f'{mission_handle.execution.identifier}.'
```
Same here, `mission_handle.activity`.

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py:
```diff
@@ -0,0 +1,767 @@
+        retry_count = 10
+        while count < retry_count and not mission_queue_id:
+            if (docking_mission_identifier !=
+                    mission_handle.execution.identifier):
```
Same here, `mission_handle.activity`.

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py:
```diff
@@ -0,0 +1,767 @@
+            self.node.get_logger().info(
+                f'[{self.name}] Failed to request robot to dock to '
+                f'destination. Maximum request navigate retries exceeded!')
+            mission_handle.execution.finished()
```
I don't think we should trigger `finished` here... that would tell RMF to continue with a task even though the right conditions aren't met.

This seems like a case where we should use the `nav_issue_ticket` and maybe trigger a replan.

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py:
```diff
@@ -0,0 +1,767 @@
+            self.node.get_logger().info(
+                f'Robot [{self.name}] is busy with another perform action! '
+                f'Ignoring new action [{category}]')
+            execution.finished()
```
Triggering `execution.finished()` here is pretty questionable. It means RMF will behave as though the newly requested action was completed even though it wasn't even performed.

It's true that this situation shouldn't generally happen, but I think if we need to consider it then we should preempt the ongoing action by simply pretending it isn't happening and performing the new action as though everything is normal. But maybe log an error message to indicate that something unusual has happened.

xiyuoh replied:
Thanks for pointing this out. It actually helped me find an edge case where this could happen in the current code. In cart delivery, if the category supplied is invalid, [`execution.finished()` is called](https://github.com/open-rmf/fleet_adapter_mir/blob/xiyu/efc_fleet_adapter_mir/fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_cart_delivery.py#L151), but the robot adapter's [`self.current_action` is created](https://github.com/open-rmf/fleet_adapter_mir/blob/xiyu/efc_fleet_adapter_mir/fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py#L756) either way.
This should not happen as category matching was performed before creating the plugin, and additionally with your suggestion for `supports_action` I believe it is safe to remove this default case and the `execution.finished()` call.

I also wonder if it would be helpful to add a `self.abandon_action` flag in MirAction. If it is called, we will return `True` in `update_action()` and the robot adapter would set its `self.current_action` to `None`.

Lastly to clarify, do you mean that in this scenario we should let `self.current_action = None` and accept the new `perform_action` received?

xiyuoh replied:
After working on the remaining refactor, I don't have very strong reasons to add an `end_action` flag in MirAction.

Regarding this unlikely scenario, I've modified the behavior to accept the new incoming actions, and end the current action in [9610e97](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/9610e9788b5ee3f9caca31d1b69b2b3ef53235a3)

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py:
```diff
@@ -0,0 +1,767 @@
+            if category in actions:
+                # Import relevant plugin
+                module = config['module']
+                plugin = importlib.import_module(module)
```
It seems that we're loading the plugin each time we're asked to perform an action. This has a few problems:
* Loading a module with importlib is a substantial operation that involves parsing and processing the Python code of the entire module that's being loaded, including all of its imports, transitively. This is an expensive operation to be doing repeatedly.
* If the plugin is not configured correctly, we won't find out until the robot is performing a task and this function crashes.

Instead we should use importlib right when the fleet adapter is starting up. Instead of passing around a `plugin_config` dictionary that contains the config file information, we should pass around an already processed dictionary that maps from the action name to the `ActionFactory`. If there are any errors in the config file, we'll be able to catch that when we find that we're not able to create the `ActionFactory`.

mxgrey commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/mir_action.py:
```diff
@@ -0,0 +1,114 @@
+                    mir_api: MirAPI,
+                    update_handle,  # rmf_fleet_adapter.RobotUpdateHandle
+                    fleet_config: rmf_easy.FleetConfiguration,
+                    action_config) -> MirAction:
```
There are too many arguments for this function. If we ever want to tweak what kind of information is passing through this API, we would end up breaking the API which will affect everyone who has integrated against it.

I recommend creating a new class called something like `ActionContext` whose only role is to store these fields. Then `MirActionFactory.make_action` should just take `self` and `context: ActionContext` as arguments. Then the constructor for `MirAction` can also take in only an `ActionContext` and store that as a single `self.context` field instead of copying every argument into self.

mxgrey commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/mir_action.py:
```diff
@@ -0,0 +1,114 @@
+    def __init__(self):
+        pass
+
+    def make_action(self,
```
This should be marked with `@abstractmethod`

mxgrey commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/mir_action.py:
```diff
@@ -0,0 +1,114 @@
+            current_task_id, [label], lambda result: _on_cancel(result))
+
+
+class MirActionFactory(ABC):
```
I suggest we also have
```
@abstractmethod
def supports_action(category: str) -> bool
```
so that at startup we can verify that every action factory really supports the categories of actions that the config file says it should.

mxgrey commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/mir_action.py:
```diff
@@ -0,0 +1,114 @@
+
+    # This will be called whenever an action has begun
+    @abstractmethod
+    def perform_action(self,
```
It's a bit clunky that the same action instance could have `perform_action` called on it multiple times. I don't think that [affordance](https://hackernoon.com/affordance-in-software-design-12cc0d9d2721) matches the intention of the API.

It might be better to merge this function with `ActionFactory.make_action`. Perhaps we could have something like the following:

```
class ActionFactory(ABC):
    ...
    def perform_action(
        self,
        category: str,
        description: dict,
        context: ActionContext
    ) -> MirAction:
```
I think `execution` could go inside of `ActionContext`.

Then `ActionFactory` would be responsible for looking at the category and description and then initializing the action accordingly. As soon as the `MirAction` is created, it begins performing.

mxgrey commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_cart_delivery.py:
```diff
@@ -0,0 +1,512 @@
+        assert (detection_module is not None,
+                'CartDetection module is required for CartDelivery plugin, ' +
+                'but it is not found!')
+        detection_plugin = importlib.import_module(detection_module)
```
Ideally this should be something parsed and loaded by the `ActionFactory` during the configuration step while the fleet adapter is starting up.

mxgrey commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_cart_detection.py:
```diff
@@ -0,0 +1,121 @@
+from urllib.error import HTTPError
+
+
+class CartDetection:
```
Seems like this should be `ABC` and all the methods should have `@abstractmethod`.

xiyuoh replied:
This makes sense, thanks for pointing it out.

xiyuoh commented:
Thanks for the meticulous review @mxgrey ! I've gone through another round of refactor, with the following changes:
- Import action plugins and any supporting modules only once at startup. Each plugin's `ActionFactory` is instantiated once, and the adapter stores these factories in a dictionary. The dictionary key is the `plugin_name` as opposed to the suggested `action` name, as each plugin/`ActionFactory` could support multiple actions.
- Using an `ActionContext` to store important handles needed throughout the plugin.
   After some thought, I decided to create the `ActionContext` at startup and exclude `execution`. This is because the `ActionFactory` may need access to some of the `ActionContext` members (e.g. access to `MirAPI` for creating missions) before a `perform_action()` is called.
- At startup, the `ActionFactory` ensures that the imported config file contains a list of supported actions, otherwise it raises an error ([55b8877](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/55b8877af7b06f678b27875cb978bba622e76357)). Depending on whether an action missions JSON filepath is provided, it will create the plugin-related mission on the robot. This mission creation step was previously inside the `MirAction.init()`, which meant that the adapter had to check for existing missions at every `perform_action()` call.
- Raise errors if plugin-specific keys are not provided in the `rmf_cart_delivery` config: [55b8877](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/55b8877af7b06f678b27875cb978bba622e76357)
- Added `ActionFactory.supports_action()` and implemented in `rmf_cart_delivery`
- Splitting `CartDelivery` into `CartPickup` and `CartDropoff`. I think it would be a better demonstration of what you mentioned about a single plugin supporting multiple actions.
- Adding cancellation behavior to `dispatch_delivery`. This ensures that if the delivery task is cancelled in the middle of a cart pickup, the robot will dropoff the cart before ending the task.
- Some fixes to the existing `CartPickup` and `CartDropoff` logic
- Minor fixes and improvements to address review comments: [a153c8d](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/a153c8dc0fece9ac39249497a39d108f9642ea06), [cecab1e](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/cecab1e49d24a92baa657b85341dc7f73a64839b), [9610e97](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/9610e9788b5ee3f9caca31d1b69b2b3ef53235a3)
- Better documentation ([30d58ff](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/30d58ffb5b0bafd9f2aed7b4b349677aee1e9821)). The links provided to the new docs in README are pointing to the `main` branch, which doesn't exist right now. Here are the newly added [`mir_missions.md`](https://github.com/open-rmf/fleet_adapter_mir/blob/xiyu/efc_fleet_adapter_mir/docs/mir_missions.md) and [`mir_actions.md`](https://github.com/open-rmf/fleet_adapter_mir/blob/xiyu/efc_fleet_adapter_mir/docs/mir_actions.md).

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/fleet_adapter_mir.py:
```diff
@@ -55,229 +68,138 @@ def sanitise_dict(dictionary, inplace=False, recursive=False):
+    conversions = config_yaml['conversions']
+    update_frequency = config_yaml['rmf_fleet']['robot_state_update_frequency']
+    debug = config_yaml['rmf_fleet']['debug']
+    plugin_config = config_yaml.get('plugins')
```
I don't see anywhere that we're calling [`add_performable_action`](https://github.com/open-rmf/rmf_ros2/blob/1db8a6559c58994f5024138daa1d04b2bec2812a/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/FleetUpdateHandle.hpp#L192-L205). If I remember correctly the fleet adapter will reject all tasks that include an action which we haven't added to the fleet, so robots will only be able to do patrols if we don't call that function.

I recommend iterating through all the plugins and adding performable actions for each action listed in the `actions:` list of each plugin.

xiyuoh replied:
I initially intended for users to duplicate their plugin actions in the fleet config to add them in as performable actions, but this is a much better solution that minimizes user diligence. Added in [49724d8](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/49724d88bc58dcfa864733867062d30aee7fc8a7)

mxgrey replied:
It occurs to me that a user might sometimes want a custom [consideration callback](https://github.com/open-rmf/rmf_ros2/blob/dd91566c7e9483bfa0d053d9ad1846b5fc725ae6/rmf_fleet_adapter/include/rmf_fleet_adapter/agv/FleetUpdateHandle.hpp#L205), in which case they'll need to call `add_performable_action` manually.

We could consider having a config parameter like `auto_add_action: False` to give the user the chance to set this manually while their factory is being built. But this isn't important for this PR, so let's just open an issue ticket for this and leave it for later if anyone wants it.

mxgrey commented on fleet_adapter_mir/fleet_adapter_mir/robot_adapter_mir.py:
```diff
@@ -0,0 +1,828 @@
+        self.current_action = None
+
+        # Import and store plugin actions and action factories
+        self.action_factories = {}
```
I recommend iterating through the `plugin_config` and creating a dictionary that maps from action category to plugin name. Then when performing an action you would look up which plugin the user specified for that action and use that specific action factory instead of looping through all of them. This would help make sure we use the action factory that the user intended in situations where the same action could be performed by multiple different plugins. If the user specifies the same action category for multiple plugins then we would throw an exception right away.

We could still fall back on iterating through the action factories in situations where an incoming action category does not match any that were listed. That could happen if the user manually calls `add_performable_action` downstream.

xiyuoh replied:
Gotcha, added the action-to-plugin dictionary in [8239b87](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/8239b87f58dbc0edad472705c1b8df7ecff89014).

> We could still fall back on iterating through the action factories in situations where an incoming action category does not match any that were listed. That could happen if the user manually calls add_performable_action downstream.

I removed the code to iterate through action factories as a fallback. I've been thinking about this and could not come up with a scenario where the user manually called `add_performable_action` downstream while the action config provides clue to which plugin this new action belongs to. If the action has already been provided in the action config, we would have added them as performable. If the action has not been provided, we would not know which plugin it is associated with.

mxgrey replied:
> If the action has not been provided, we would not know which plugin it is associated with.

The idea is that if a user manually calls `add_performable_action`, we can figure out which plugin to use by iterating through each factory and calling `factory.supports_action( _ )` to see which plugin supports the action.

That being said, this would be an extremely niche use case to the point that I wouldn't mind if we officially don't support it until someone asks for it with a compelling enough use case. It should be possible to add support for it in the future without breaking API, so we don't need to worry about this hypothetical right now.

mxgrey replied:
If you want to add these capabilities within this PR I have no objection, but I'm also perfectly happy to merge as-is and just leave an issue ticket for us to consider those cases in the future.

mxgrey reviewed (changes requested):
This PR is looking very good and getting super super close.

I left just a few more items of feedback which I think should be pretty straightforward to address. After those are done we should be fully ready to merge.

xiyuoh commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/mir_action.py:
```diff
@@ -0,0 +1,126 @@
+                            cancel_success: Callable[[], None],
+                            cancel_fail: Callable[[], None],
+                            label: str = ''):
+        current_task_id = self.context.update_handle.more().current_task_id()
```
:+1: [b87a704](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/b87a704f7d62557aa4ee6ff6e2b5ffb74dac0462)

xiyuoh commented on fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_cart_delivery.py:
```diff
@@ -0,0 +1,484 @@
+        detection_plugin = importlib.import_module(detection_module)
+        self.cart_detection = detection_plugin.CartDetection(self.context)
+
+        # Verify that this ActionFactory supports all the actions specified in
```
👍🏻 [a7aac27](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/a7aac274df3e771111d5dc58ed3dc87a22fc8d58)

mxgrey reviewed (approved):
Awesome, thanks for all the massive iterating.

This is without question the most professional and modular RMF fleet adapter known to exist. As we work on the next generation of Open-RMF's SDK we'll definitely want to take advantage of lessons learned in this effort to figure out how to lift as much of this upstream as we can.

Great job bringing this to completion :tada:

xiyuoh commented:
Thanks Grey, really appreciate the detailed review! I've updated with the changes mentioned since they're small and quick to add in. I'll have to trouble you to look at [45ac09a](https://github.com/open-rmf/fleet_adapter_mir/pull/36/commits/45ac09adba01be6752192c903f739067916aae61) that I just pushed, it happens to remove the exception message with the grammar mistake.
